src/glyph_item_iter.rs

```diff
@@ -0,0 +1,229 @@
+        free => |ptr| pango_sys::pango_glyph_item_iter_free(ptr),
+        init => |_ptr| (),
+        clear => |_ptr| (),
+        get_type => || pango_sys::pango_glyph_item_iter_get_type(),
+    }
+}
+
+impl GlyphItemIter {
+    pub fn init_end(glyph_item: &GlyphItem, text: &str) -> Option<GlyphItemIter> {
+        unsafe {
+            let mut iter = GlyphItemIter::uninitialized();
+            let ret = from_glib(pango_sys::pango_glyph_item_iter_init_end(
+                iter.to_glib_none_mut().0,
+                mut_override(glyph_item.to_glib_none().0),
+                text.to_glib_none().0,
+            ));
+            if ret {
+                Some(iter)
+            } else {
+                None
+            }
+        }
+    }
+
+    pub fn init_start(glyph_item: &GlyphItem, text: &str) -> Option<GlyphItemIter> {
+        unsafe {
+            let mut iter = GlyphItemIter::uninitialized();
+            let ret = from_glib(pango_sys::pango_glyph_item_iter_init_start(
+                iter.to_glib_none_mut().0,
+                mut_override(glyph_item.to_glib_none().0),
+                text.to_glib_none().0,
+            ));
+            if ret {
+                Some(iter)
+            } else {
+                None
+            }
+        }
+    }
+
+    pub fn next_cluster(&mut self) -> bool {
+        unsafe {
+            from_glib(pango_sys::pango_glyph_item_iter_next_cluster(
+                self.to_glib_none_mut().0,
+            ))
+        }
+    }
+
+    pub fn prev_cluster(&mut self) -> bool {
+        unsafe {
+            from_glib(pango_sys::pango_glyph_item_iter_prev_cluster(
+                self.to_glib_none_mut().0,
+            ))
+        }
+    }
+
+    pub fn start_index(&self) -> i32 {
+        self.0.start_index
+    }
+
+    pub fn start_char(&self) -> i32 {
+        self.0.start_char
+    }
+
+    pub fn start_glyph(&self) -> i32 {
+        self.0.start_glyph
+    }
+
+    pub fn end_index(&self) -> i32 {
+        self.0.end_index
+    }
+
+    pub fn end_char(&self) -> i32 {
+        self.0.end_char
+    }
+
+    pub fn end_glyph(&self) -> i32 {
+        self.0.end_glyph
+    }
+
+    pub fn into_data(&self) -> GlyphItemIteratorData {
+        GlyphItemIteratorData {
+            start_glyph: self.0.start_glyph,
+            end_glyph: self.0.end_glyph,
+            start_index: self.0.start_index as usize,
+            end_index: self.0.end_index as usize,
+            start_char: self.0.start_char as usize,
+            end_char: self.0.end_char as usize,
+        }
+    }
+}
+
+pub struct GlyphItemIteratorData {
+    pub start_glyph: i32,
+    pub start_index: usize,
+    pub start_char: usize,
+
+    pub end_glyph: i32,
+    pub end_index: usize,
+    pub end_char: usize,
+}
+
+pub struct GlyphItemIterator<'a> {
+    item: &'a GlyphItem,
+    text: &'a str,
+    is_reverse: bool,
+    iter: Option<GlyphItemIter>,
+}
+
+impl<'a> GlyphItemIterator<'a> {
+    #[inline(always)]

```

Comment:
Why?

Reply:
Hm, this function and next unneeded

Reply:
I meant the inline annotation. Don't add those unless you benchmarked that it actually makes a positive difference :)

Reply:
Updated, also removed old unused getters

Reply:
I would argue that it's a good idea to mark it `#[inline]`, because otherwise functions without type parameter like this would not be inlined downstream unless the final binary is built with LTO enabled.

Reply:
Does it even make a difference if it's not inlined?

Reply:
It would have an extra function call whenever downstream invokes it, and if that happens to be in a tight loop, it could be a problem.

My thought is basically that marking it inline wouldn't hurt anything as it's so small. It would just ensure the compiler is able to inline it if it thinks fit.

But maybe applications serious on performance should really enable LTO, so it may not really matter that much.

Reply:
In theory I agree with you but in practice it's all not that easy and generally a better idea to not do anything unless you actually measure it. The practical effects are not always that obvious. Just like with all optimizations: measure, then fix if needed.